File: main/java/com/websarva/wings/android/dowasurememo/CarMemo.java
package com.websarva.wings.android.dowasurememo;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;

public class CarMemo extends AppCompatActivity {

    private static final String _CATEGORY = "CAR";

    /**
     * データベースのテーブル名
     */
    private static final String TABLE="car";

    private Context context=CarMemo.this;

    private LayoutInflater inflater;
    private LinearLayout linearLayout;
    private LinearLayout llCarLayout;
    private LinearLayout llCarNameInputform;
    private LinearLayout llCarDetailInputform;

    private EditText etCarName;
    private EditText etCarMemoTitle;
    private EditText etCarMemoContents;
    private ImageButton btDelete;
    private ImageButton btCarDetailAdd;

    private String strCarName;
    private String strCarMemoTitle;
    private String strCarMemoContents;

    private String name="name";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_car_memo);

        llCarLayout=findViewById(R.id.ll_car_layout);

        inflater = LayoutInflater.from(getApplicationContext());
        llCarLayout=findViewById(R.id.ll_car_layout);

        /**
         * データベースの列名の配列
         */
        String[] columnNames={"carname","inputform"};

        String inputform="name";

        //データベースからデータを取り出して、レイアウトを作成する処理
        DatabaseControl control=new DatabaseControl(context,TABLE,columnNames);
        control.selectDatabase(llCarLayout);



        /**
         * データベースの列名の配列
         */
        String[] columnNames2={"carmemotitle","carmemocontents","inputform"};



        DatabaseControl control2=new DatabaseControl(context,TABLE,columnNames2);
        control2.selectDatabase(llCarLayout);


//        //データベースからデータを取り出して、viewを生成する処理
//        _helper=new Databasehelper(getApplicationContext());
//        SQLiteDatabase db=_helper.getWritableDatabase();
//        String sqlSelect="SELECT * FROM car";
//        Cursor cursor=db.rawQuery(sqlSelect,null);
//        cursor.moveToFirst();
//        while (cursor.moveToNext()) {
//
//            int i = cursor.getColumnIndex("_id");
//            tagId = cursor.getInt(i);
//            i=cursor.getColumnIndex("inputform");
//            String st=cursor.getString(i);
//            if(st.equals(name)){
//                //車の車種を入力するviewを生成
//                inflater = LayoutInflater.from(getApplicationContext());
//                llCarLayout=findViewById(R.id.ll_car_layout);
//                llCarNameInputform=(LinearLayout)inflater.inflate(R.layout.car_name_inputform,null);
//                llCarLayout.addView(llCarNameInputform);
//
//                //viewの削除ボタンにリスナーをセット
//                btDelete=llCarNameInputform.findViewById(R.id.bt_delete);
//                btDelete.setOnClickListener
//                        (new DeleteButton(CarMemo.this,llCarLayout,llCarNameInputform,table));
//
//                //viewの追加ボタンにリスナーをセット
//                btCarDetailAdd=llCarNameInputform.findViewById(R.id.bt_cardetail_add);
//                btCarDetailAdd.setOnClickListener(new AddCarDetail(CarMemo.this));
//
//                etCarName=llCarNameInputform.findViewById(R.id.et_car_name);
//
//                i = cursor.getColumnIndex("carname");
//                strCarName = cursor.getString(i);
//
//                try {
//                    etCarName.setText(strCarName);
//                } catch (NullPointerException e) {
//                    strCarName = "";
//                }
//
//            }
//            else{
//                //車のナンバーなどの詳細を入力するviewを生成
//                inflater = LayoutInflater.from(getApplicationContext());
//                llCarLayout=findViewById(R.id.ll_car_layout);
//                llCarDetailInputform= (LinearLayout) inflater.inflate(R.layout.car_detail_inputform,null);
//                llCarLayout.addView(llCarDetailInputform);
//
//                //viewの削除ボタンにリスナーをセット
//                btDelete=llCarDetailInputform.findViewById(R.id.bt_delete);
//                btDelete.setOnClickListener
//                        (new DeleteButton(CarMemo.this,llCarLayout,llCarDetailInputform,table));
//
//
//                etCarMemoTitle=llCarDetailInputform.findViewById(R.id.et_car_memo_title);
//                etCarMemoContents=llCarDetailInputform.findViewById(R.id.et_car_memo_contents);
//
//                i = cursor.getColumnIndex("carmemotitle");
//                strCarMemoTitle = cursor.getString(i);
//
//                i = cursor.getColumnIndex("carmemocontents");
//                strCarMemoContents = cursor.getString(i);
//
//                try {
//                    etCarMemoTitle.setText(strCarMemoTitle);
//                } catch (NullPointerException e) {
//                    strCarMemoTitle = "";
//                }
//
//                try {
//                    etCarMemoContents.setText(strCarMemoContents);
//                } catch (NullPointerException e) {
//                    strCarMemoContents = "";
//                }
//            }
//        }
//
//        /*なぜか最初にviewを追加すると、１番目に追加したviewだけ消えてしまう事象が発生したため
//        コード上であらかじめ追加しておいた1番目のviewを非表示にする処理*/
//        if(llCarLayout.getChildCount()!=0){
//            LinearLayout firstView= (LinearLayout) llCarLayout.getChildAt(0);
//            firstView.setVisibility(View.GONE);
//        }else {
//        }

    }

    //車の詳細を入力するviewを追加するための処理
    public class AddCarDetail extends LinearLayout implements View.OnClickListener{
        public AddCarDetail(Context context) {
            super(context);
        }
        @Override
        public void onClick(View v) {
            llCarDetailInputform= (LinearLayout) inflater.inflate(R.layout.car_detail_inputform,null);
            linearLayout= (LinearLayout) v.getParent();
            for(int i=0;i<llCarLayout.getChildCount();i++){
                LinearLayout sameLinearLayout= (LinearLayout) llCarLayout.getChildAt(i);
                if(linearLayout==sameLinearLayout){
                    llCarLayout.addView(llCarDetailInputform,i+1);

                    btDelete=llCarDetailInputform.findViewById(R.id.bt_delete);
                    btDelete.setOnClickListener
                            (new DeleteButton(CarMemo.this,llCarLayout,llCarDetailInputform,TABLE));

                    etCarMemoTitle=llCarDetailInputform.findViewById(R.id.et_car_memo_title);
                    etCarMemoContents=llCarDetailInputform.findViewById(R.id.et_car_memo_contents);

                }
            }
        }
    }

    //オプションメニューの実装
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater=getMenuInflater();
        inflater.inflate(R.menu.optionmenu,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    @Override
    //車の車種を入力するviewを追加するための処理
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {

        switch (item.getItemId()) {
            case R.id.option_add:
//                /*なぜか最初にviewを追加すると、１番目に追加したviewだけ消えてしまう事象が発生したため
//                コード上であらかじめ1番目のviewを追加して非表示にする処理*/
//                if(llCarLayout.getChildCount()==0){
//                    inflater = LayoutInflater.from(getApplicationContext());
//                    llCarNameInputform=(LinearLayout)inflater.inflate(R.layout.car_name_inputform,null);
//                    llCarLayout.addView(llCarNameInputform);
//                    llCarNameInputform.setVisibility(View.GONE);
//
//                    String str="";
//                    DatabaseControl control=new DatabaseControl(context,table);
//                    control.deleteDatabase(1);
//
//                    String column1="carname";
//                    String column2="inputform";
//                    String inputform="name";
//
//                    DatabaseControl control2=new DatabaseControl
//                            (context,table,1,_category,str,inputform);
//                    control2.insertDatabaseTwoColumns(column1,column2);
//                }

                //オプションメニューの＋ボタンを押すと、動的にビューを追加する処理
                inflater = LayoutInflater.from(getApplicationContext());
                llCarNameInputform=(LinearLayout)inflater.inflate(R.layout.car_name_inputform,null);
                llCarLayout.addView(llCarNameInputform);

                btCarDetailAdd=llCarNameInputform.findViewById(R.id.bt_cardetail_add);
                btCarDetailAdd.setOnClickListener(new AddCarDetail(CarMemo.this));

                btDelete=llCarNameInputform.findViewById(R.id.bt_delete);
                btDelete.setOnClickListener
                        (new DeleteButton(CarMemo.this,llCarLayout,llCarNameInputform,TABLE));

        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPause() {
        super.onPause();

        //データベースにある全てのデータを削除
        DatabaseControl control = new DatabaseControl(context, TABLE);
        control.deleteAllDatabase();

        //メモの文字列を取得してデータベースにインサートする
        for(int i=0;i<llCarLayout.getChildCount();i++){
            LinearLayout linearLayout= (LinearLayout) llCarLayout.getChildAt(i);
            switch (linearLayout.getId()){
                case R.id.ll_car_name_inputform:
                    etCarName=linearLayout.findViewById(R.id.et_car_name);
                    strCarName=etCarName.getText().toString();

                    String inputform="name";

                    DatabaseControl control2=new DatabaseControl
                            (context,TABLE, i, _CATEGORY,strCarName,inputform);
                    control2.insertDatabaseTwoColumns("carname","inputform");

                    break;



                case R.id.ll_car_detail_inputform:
                    etCarMemoTitle=linearLayout.findViewById(R.id.et_car_memo_title);
                    etCarMemoContents=linearLayout.findViewById(R.id.et_car_memo_contents);
                    strCarMemoTitle=etCarMemoTitle.getText().toString();
                    strCarMemoContents=etCarMemoContents.getText().toString();

                    String memoinputform="detail";

                    DatabaseControl control3=new DatabaseControl
                            (context,TABLE, i, _CATEGORY,strCarMemoTitle,strCarMemoContents,memoinputform);
                    control3.insertDatabaseThreeColumns("carmemotitle","carmemocontents","inputform");

                    break;


            }
        }
    }
}

File: main/java/com/websarva/wings/android/dowasurememo/CarMemo.kt
package com.websarva.wings.android.dowasurememo

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.*
import android.widget.ImageButton
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import com.websarva.wings.android.dowasurememo.databinding.*

class CarMemo : AppCompatActivity() {

    companion object {
        private const val _CATEGORY = "CAR"

        /**
         * データベースのテーブル名
         */
        private const val TABLE = "car"
    }

    private val context: Context = this@CarMemo
//    private lateinit var inflater: LayoutInflater

//    private var llCarLayout: LinearLayout? = null
//    private var llCarNameInputform: LinearLayout? = null
//    private var etCarName: EditText? = null
//    private var etCarMemoTitle: EditText? = null
//    private var etCarMemoContents: EditText? = null
//    private var btDelete: ImageButton? = null
//    private var btCarDetailAdd: ImageButton? = null
//    private var strCarName: String? = null
//    private var strCarMemoTitle: String? = null
//    private var strCarMemoContents: String? = null

    private lateinit var binding: ActivityCarMemoBinding
    private lateinit var carNameBinding: CarNameInputformBinding
    private lateinit var carDetailBinding: CarDetailInputformBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCarMemoBinding.inflate(layoutInflater)
        carNameBinding = CarNameInputformBinding.inflate(layoutInflater)
        carDetailBinding = CarDetailInputformBinding.inflate(layoutInflater)
        setContentView(binding.root)

//        llCarLayout = findViewById(R.id.ll_car_layout)
//        inflater = LayoutInflater.from(applicationContext)
//        llCarLayout = findViewById(R.id.ll_car_layout)

        //データベースからデータを取り出して、レイアウトを作成する処理
        DatabaseControl(context, TABLE).selectDatabase(binding.llCarLayout)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.optionmenu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    @RequiresApi(api = Build.VERSION_CODES.Q)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.option_add -> {
                //オプションメニューの＋ボタンを押すと、動的にビューを追加する処理
//                inflater = LayoutInflater.from(applicationContext)
                //llCarNameInputform = inflater?.inflate(R.layout.car_name_inputform, null) as LinearLayout
                binding.llCarLayout.addView(carNameBinding.llCarNameInputform)
//                btCarDetailAdd = llCarNameInputform!!.findViewById(R.id.bt_cardetail_add)
//                btCarDetailAdd?.setOnClickListener(AddCarDetail(this@CarMemo, llCarLayout))
//                btDelete = llCarNameInputform!!.findViewById(R.id.bt_delete)
                carNameBinding.btDelete.setOnClickListener(DeleteButton(this@CarMemo, binding.llCarLayout, carNameBinding.llCarNameInputform, TABLE))
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onPause() {
        super.onPause()

        //データベースにある全てのデータを削除
        DatabaseControl(context, TABLE).deleteAllDatabase()

        //メモの文字列を取得してデータベースにインサートする
        for (i in 0..binding.llCarLayout.childCount) {
            val linearLayout = binding.llCarLayout.getChildAt(i) as LinearLayout
            when (linearLayout.id) {
                R.id.ll_car_name_inputform -> {
                    //carNameBinding.etCarName = linearLayout.findViewById(R.id.et_car_name)
                    var strCarName = carNameBinding.etCarName.text.toString()
                    val inputform = "name"
                    DatabaseControl(context, TABLE, i, _CATEGORY, strCarName, inputform)
                            .insertDatabaseTwoColumns("carname", "inputform")
                }
                R.id.ll_car_detail_inputform -> {
//                    etCarMemoTitle = linearLayout.findViewById(R.id.et_car_memo_title)
//                    etCarMemoContents = linearLayout.findViewById(R.id.et_car_memo_contents)
                    var strCarMemoTitle = carDetailBinding.etCarMemoTitle.text.toString()
                    var strCarMemoContents = carDetailBinding.etCarMemoContents.text.toString()
                    val memoinputform = "detail"
                    DatabaseControl(context, TABLE, i, _CATEGORY, strCarMemoTitle, strCarMemoContents, memoinputform)
                            .insertDatabaseThreeColumns("carmemotitle", "carmemocontents", "inputform")
                }
            }
        }
        finish()
    }


}

/**
 * 車の詳細を入力するviewを追加するためのクラス
 *
 * @author nakayama
 * @version 1.0
 */
internal class AddCarDetail(context: Context, private val llBaseLayout: LinearLayout?) : LinearLayout(context), View.OnClickListener {

    private lateinit var inflater: LayoutInflater
    private var llCarDetailInputform: LinearLayout? = null
    private val TABLE = "car"

    override fun onClick(v: View) {
        /*押されたボタンの親ビューとllBaseLayoutの子ビューが等しいときに、
        そのビューのインデックス＋１した場所に車の詳細を入力するビューを追加する*/
        inflater = LayoutInflater.from(context)
        llCarDetailInputform = inflater.inflate(R.layout.car_detail_inputform, null) as LinearLayout
        val linearLayout = v.parent as LinearLayout
        for (i in 0..llBaseLayout!!.childCount) {
            val sameLinearLayout = llBaseLayout!!.getChildAt(i) as LinearLayout
            if (linearLayout === sameLinearLayout) {
                llBaseLayout.addView(llCarDetailInputform, i + 1)
                val btDelete = llCarDetailInputform!!.findViewById<ImageButton>(R.id.bt_delete)
                btDelete.setOnClickListener(DeleteButton(context, llBaseLayout, llCarDetailInputform, TABLE))
            }
        }
    }


}